cpp/ql/src/experimental/Security/CWE/CWE-409-DecompressionBomb/DecompressionBombs.ql

```diff
@@ -0,0 +1,186 @@
+      fc.getArgument(2) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof Archive_read_data_block |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof MzUncompress |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof MzZipReaderExtract |
+      fc.getArgument(1) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof MzInflate |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof TinflDecompress |
+      fc.getArgument(1) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof TinflDecompressMem |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof ZSTDDecompressFunction |
+      fc.getArgument(2) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof ZSTDDecompressDCtxFunction |
+      fc.getArgument(3) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof ZSTDDecompressStreamFunction |
+      fc.getArgument(2) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof ZSTDDecompressUsingDictFunction |
+      fc.getArgument(3) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof ZSTDDecompressUsingDDictFunction |
+      fc.getArgument(3) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof UnzReadCurrentFileFunction |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof Mz_zip_reader_entry |
+      fc.getArgument(1) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof Mz_zip_entry |
+      fc.getArgument(1) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof LzmaStreamBufferDecodeFunction |
+      fc.getArgument(1) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof LzmaCodeFunction |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof GzReadFunction |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof GzFreadFunction |
+      sink.asExpr() = fc.getArgument(3)
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof GzGetsFunction |
+      sink.asExpr() = fc.getArgument(0)
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof InflateFunction |
+      fc.getArgument(0) = sink.asExpr()
+    )
+    or
+    exists(FunctionCall fc | fc.getTarget() instanceof UncompressFunction |
+      fc.getArgument(0) = sink.asExpr()
+    )
+  }
+
+  predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
```

Comment:
Similar to `isSink` this can probably be made more compact with the help of an abstract class.